File: server/internal/domain/evaluation/contract.go
package evaluation

import (
	"context"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"

	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/backtest"
)

type EvaluationInput struct {
	RunID         uuid.UUID
	InitialEquity decimal.Decimal
	Trades        []backtest.TradeFact
	EquityPoints  []backtest.EquityPoint
}

type EvaluationPolicy struct {
	EvaluatorVersion    string
	PeriodsPerYear      int
	ZeroPnLCountsAsWin  bool
	StddevDDOF          int
	MinPeriodsForSharpe int
	RiskFreeRate        decimal.Decimal
}

type Evaluation struct {
	BacktestRunID    uuid.UUID        `json:"backtest_run_id"`
	EvaluatorVersion string           `json:"evaluator_version"`
	TotalReturnPct   decimal.Decimal  `json:"total_return_pct"`
	WinRatePct       decimal.Decimal  `json:"win_rate_pct"`
	MaxDrawdownPct   decimal.Decimal  `json:"max_drawdown_pct"`
	TradeCount       int              `json:"trade_count"`
	OpenTradeCount   int              `json:"open_trade_count"`
	ProfitFactor     *decimal.Decimal `json:"profit_factor,omitempty"`
	SharpeRatio      *decimal.Decimal `json:"sharpe_ratio,omitempty"`
	AvgTradePct      *decimal.Decimal `json:"avg_trade_pct,omitempty"`
}

type Evaluator interface {
	Evaluate(context.Context, EvaluationInput, EvaluationPolicy) (Evaluation, error)
}

File: py_backend/app/domain/evaluation/contract.py
"""Evaluation contracts (float64).

Mirrors `server/internal/domain/evaluation/contract.go`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol
from uuid import UUID

from ..backtest import EquityPoint, TradeFact
from ..common import Decimal


@dataclass
class EvaluationInput:
    run_id: UUID
    initial_equity: Decimal
    trades: list[TradeFact] = field(default_factory=list)
    equity_points: list[EquityPoint] = field(default_factory=list)


@dataclass
class EvaluationPolicy:
    evaluator_version: str = ""
    periods_per_year: int = 0
    zero_pnl_counts_as_win: bool = False
    stddev_ddof: int = 1
    min_periods_for_sharpe: int = 0
    risk_free_rate: Decimal = 0.0


@dataclass
class Evaluation:
    backtest_run_id: UUID
    evaluator_version: str
    total_return_pct: Decimal
    win_rate_pct: Decimal
    max_drawdown_pct: Decimal
    trade_count: int
    open_trade_count: int
    profit_factor: Decimal | None = None
    sharpe_ratio: Decimal | None = None
    avg_trade_pct: Decimal | None = None


class Evaluator(Protocol):
    def evaluate(self, input_: EvaluationInput, policy: EvaluationPolicy) -> Evaluation: ...
